Fix stm32 dma callback and busy flag handling

Richy42 opened the pull request:
stm32: dma: fix incorrect busy flag  reset and callback check

JarmouniA reviewed (dismissed):
Your commit message title says `drivers: pwm:.. ` but the changes are in dma.

JarmouniA commented:
Cc @AUtech-dev as this is changing code added in #98057.

Richy42 committed:
drivers: dma: stm32: fix check for callback

The check for an available callback was incorrect and
the interrupt was enabled if no callback is present
while in non-cyclic mode.

Signed-off-by: Richard Skriwanek <richy@fnc.at>

Richy42 committed:
drivers: dma: stm32: fix busy flag not reset correctly

Reset the busy flag also if the function stm32_dma_is_enabled_stream()
shows a disabled stream, as the HW flag that that function reads
is automatically disabled in non-cyclic mode while the
flag busy flag is still true.

Signed-off-by: Richard Skriwanek <richy@fnc.at>

Richy42 commented on drivers/dma/dma_stm32.c:
```diff
@@ -501,7 +501,7 @@ DMA_STM32_EXPORT_API int dma_stm32_configure(const struct device *dev,
 
 	/* Enable transfer complete ISR if in non-cyclic mode or a callback is requested */
-	if (!stream->cyclic || stream->dma_callback != NULL) {
+	if (!stream->cyclic && stream->dma_callback != NULL) {
```
I understand what you mean.
I created the change as my application crashed due to the missing callback and changing the check solved it.
Seems to need closer investigation on my side what happened, probably I did not have the latest code.

Richy42 replied:
As https://github.com/zephyrproject-rtos/zephyr/pull/97475 was merged only 4 days ago, I have overseen it and had it not pulled.

That fix is superior for sure and I will happily close this PR.
